### PR TITLE
release-19.2: changefeedccl: add scan boundaries based on change in set of columns

### LINKED ISSUE #41961: changefeedccl: row updates being emitting at the wrong timestamp after a column drop
Currently, the changefeed `poller` emits row updates at the `ModificationTime` of the table descriptor where the column backfill mutation has been completely applied. If the schema change was started when the table descriptor was at version 1, the table descriptor version at which the changfeed level backfill row updates are emitted would be version 4.

The above is correct in case of `ADD COLUMN` since descriptor version 4 is the first table descriptor version such that the new column is logically available. However, in case of a `DROP COLUMN`, the dropped column would logically get dropped way before this (at version 2 if the schema change started on version 1).

We should be emitting backfill row updates for dropped columns at the `ModificationTime` of the table descriptor at which they're logically dropped (which is version 2 in the above example).

### PR DESCRIPTION
**Pull request by aayushshah15**
Backport 1/1 commits from #42053.

/cc @cockroachdb/release

---

Currently, the changefeed poller detects a scan boundary when 
detects that the last version of a table descriptor has a 
mutation but the current version doesn't. In case of an `ALTER TABLE
DROP COLUMN` statement, the point at which this happens is the point
at which the schema change backfill completes. This is incorrect since 
the dropped column is logically dropped before this point.

This PR corrects this problem by instead checking that the last version of the descriptor has a mutation AND that the number of columns in the current table descriptor is different from the number of columns in the last table descriptor.

Fixes #41961

Release note (bug fix): Changefeeds now emit backfill row updates for
                            dropped column when the table descriptor drops
                            that column.

### CODE REVIEW

**Comment by danhhz** [comment]
@ajwerner what was our decision here? hold off on merging this in 19.2.2 and get it into 19.2.3?

**Comment by ajwerner** [comment]
Totally dropped the ball on this one, looking now.

**Comment by danhhz** [comment]
You're fine, I'm also a reviewer. I was intentionally waiting because I thought we were going to hold off on merging this, but dropped the ball on putting that fact down here

**Review by ajwerner — COMMENTED** [review]
What's going on w.r.t. the build here?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42669#-:-Lv7BZDo-98Ys4W3sOBx:bhvugwb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @danhhz)

**Comment by danhhz** [comment]
`GitHub API token environment variable GITHUB_API_TOKEN is not set` I think this happened to be opened while that was broken. I think it's since been fixed. I'll ping rerun

**Review by ajwerner — APPROVED** [review]
:lgtm: 

I'm cool with this change even going in to 19.2.2 if there's no drama getting it to build. This change is certainly a behavior change and seems like the riskiest of the backports we plan to actually do but it's also a good change. 

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42669#-:-Lv7EpVN7gXikDL3EkbI:b-x5p1se)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15 and @danhhz)

___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42669#-Lv7EJFb2h2OupXYBevZ:-Lv7EJFb2h2OupXYBev_:b-dea2m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9083eef01e5c6f05755fa7c3b4cded1a624f48d7/pkg/ccl/changefeedccl/cdctest/testfeed.go#L528)):*
> ```Go
> }
> 
> // ReformatJSON marshals a golang stdlib based JSON into a byte slice preserving
> ```

nit: this comment could use another pass.

**Comment by aayushshah15** [comment]
`pkg/ccl/changefeedccl/changefeed_test.go:781:9: it.SeekGE undefined (type engine.SimpleIterator has no field or method SeekGE)`
seems like there was an interface change that wasn't backported.

**Comment by petermattis** [comment]
> pkg/ccl/changefeedccl/changefeed_test.go:781:9: it.SeekGE undefined (type engine.SimpleIterator has no field or method SeekGE)
> seems like there was an interface change that wasn't backported.

Correct. See https://github.com/cockroachdb/cockroach/pull/42390. We don't want to backport the interface change. Instead, you should adapt the backport to the old interface. That is, `s/SeekGE/Seek/g`.

**Comment by ajwerner** [comment]
> `pkg/ccl/changefeedccl/changefeed_test.go:781:9: it.SeekGE undefined (type engine.SimpleIterator has no field or method SeekGE)`
> seems like there was an interface change that wasn't backported.

I want to say this was just a name change. See `Seek`?

**Review by aayushshah15 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42669)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner and @danhhz)

___
*[pkg/ccl/changefeedccl/cdctest/testfeed.go, line 528 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42669#-Lv7EJFb2h2OupXYBevZ:-Lv7Mb6qBNbcBh8w2YLn:b-axk2ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/9083eef01e5c6f05755fa7c3b4cded1a624f48d7/pkg/ccl/changefeedccl/cdctest/testfeed.go#L528)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: this comment could use another pass.
</blockquote></details>

Done. let me know if it's still unclear.

**Comment by aayushshah15** [comment]
The build should succeed now. are we good to merge this once it's green?

**Comment by ajwerner** [comment]
> The build should succeed now. are we good to merge this once it's green?

Fine by me.

**Pull request merged**
